Linux Compatibility

Check if _variant_ works on _linux_. If not, apply the necessary changes. Windows users can use _variant_ on Linux through Docker.

Tried to run on Ubuntu (linux-docker-android-20.04)

```
Cloning into 'variants'...
/bin/bash: swift: command not found
make: *** [Makefile:23: variants] Error 127
```

--- a/.github/workflows/branch.yml
+++ b/.github/workflows/branch.yml
@@ -26,7 +26,6 @@ jobs:
       if: matrix.os != 'macos-latest'
       with:
         ruby-version: 2.7 # Not needed with a .ruby-version file
-        bundler-cache: true # runs 'bundle install' and caches installed gems automatically
 
     - name: Test CLI Package
       run: swift test --enable-test-discovery

--- a/.github/workflows/branch.yml
+++ b/.github/workflows/branch.yml
@@ -22,6 +22,7 @@ jobs:
       with:
         swift-version: ${{ matrix.swift }}
     - uses: ruby/setup-ruby@v1
+      if: matrix.os != 'macos-latest' # xcode available on macos only
       with:
         ruby-version: 2.6 # Not needed with a .ruby-version file
         bundler-cache: true # runs 'bundle install' and caches installed gems automatically

--- a/.github/workflows/branch.yml
+++ b/.github/workflows/branch.yml
@@ -22,9 +22,9 @@ jobs:
       with:
         swift-version: ${{ matrix.swift }}
     - uses: ruby/setup-ruby@v1
-      if: matrix.os != 'macos-latest' # xcode available on macos only
+      if: matrix.os != 'macos-latest'
       with:
-        ruby-version: 2.6 # Not needed with a .ruby-version file
+        ruby-version: 2.7 # Not needed with a .ruby-version file
         bundler-cache: true # runs 'bundle install' and caches installed gems automatically
 
     - name: Test CLI Package

--- a/.github/workflows/branch.yml
+++ b/.github/workflows/branch.yml
@@ -19,6 +19,7 @@ jobs:
     steps:
     - uses: actions/checkout@v2
     - uses: fwal/setup-swift@v1
+      if: matrix.os != 'macos-latest'
       with:
         swift-version: ${{ matrix.swift }}
     - uses: ruby/setup-ruby@v1

--- a/.github/workflows/branch.yml
+++ b/.github/workflows/branch.yml
@@ -13,7 +13,7 @@ jobs:
     strategy:
       matrix:
         os: [macos-latest, ubuntu-latest]
-        swift: ["5.3"]
+        swift: ["5.5"]
 
     steps:
     - uses: actions/checkout@v2

--- a/.github/workflows/branch.yml
+++ b/.github/workflows/branch.yml
@@ -8,7 +8,7 @@ on:
       - develop
 jobs:
   tests:
-    name: Test with ${{ matrix.swift }} on ${{ matrix.os }}
+    name: Test with Swift ${{ matrix.swift }} on ${{ matrix.os }}
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
@@ -40,10 +40,11 @@ jobs:
       run: bundle install
       
     - name: Code coverage report
-      if: matrix.os == 'macos-latest' # xcode available on macos only
+      if: matrix.os == 'macos-latest' # on macos only
       run: bundle exec slather coverage
 
     - uses: codecov/codecov-action@v1
+      if: matrix.os == 'macos-latest' # on macos only
       with:
         token: ${{ secrets.CODECOV_TOKEN }}
         file: ./test-reports/cobertura.xml

--- a/.github/workflows/branch.yml
+++ b/.github/workflows/branch.yml
@@ -23,7 +23,7 @@ jobs:
         swift-version: ${{ matrix.swift }}
 
     - name: Test CLI Package
-      run: swift test
+      run: swift test --enable-test-discovery
 
     - name: Make install
       run: make install

--- a/.github/workflows/branch.yml
+++ b/.github/workflows/branch.yml
@@ -11,7 +11,6 @@ jobs:
     name: Test with ${{ matrix.swift }} on ${{ matrix.os }}
     runs-on: ${{ matrix.os }}
     strategy:
-      max-parallel: 1
       matrix:
         os: [macos-latest, ubuntu-latest]
         swift: ["5.5"]
@@ -41,6 +40,7 @@ jobs:
       run: bundle install
       
     - name: Code coverage report
+      if: matrix.os == 'macos-latest' # xcode available on macos only
       run: bundle exec slather coverage
 
     - uses: codecov/codecov-action@v1

--- a/.github/workflows/branch.yml
+++ b/.github/workflows/branch.yml
@@ -8,10 +8,19 @@ on:
       - develop
 jobs:
   tests:
-    runs-on: macos-latest
+    name: Test with ${{ matrix.swift }} on ${{ matrix.os }}
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [macos-latest, ubuntu-latest]
+        swift: ["5.3"]
 
     steps:
     - uses: actions/checkout@v2
+
+    - uses: fwal/setup-swift@v1
+      with:
+        swift-version: ${{ matrix.swift }}
 
     - name: Test CLI Package
       run: swift test
@@ -20,6 +29,7 @@ jobs:
       run: make install
       
     - name: Test VariantsCore
+      if: matrix.os == 'macos-latest' # xcode available on macos only
       run: xcodebuild test -scheme VariantsCore
 
     - name: Bundle install

--- a/.github/workflows/branch.yml
+++ b/.github/workflows/branch.yml
@@ -22,6 +22,9 @@ jobs:
       with:
         swift-version: ${{ matrix.swift }}
     - uses: ruby/setup-ruby@v1
+      with:
+        ruby-version: 2.6 # Not needed with a .ruby-version file
+        bundler-cache: true # runs 'bundle install' and caches installed gems automatically
 
     - name: Test CLI Package
       run: swift test --enable-test-discovery

--- a/.github/workflows/branch.yml
+++ b/.github/workflows/branch.yml
@@ -11,16 +11,17 @@ jobs:
     name: Test with ${{ matrix.swift }} on ${{ matrix.os }}
     runs-on: ${{ matrix.os }}
     strategy:
+      max-parallel: 1
       matrix:
         os: [macos-latest, ubuntu-latest]
         swift: ["5.5"]
 
     steps:
     - uses: actions/checkout@v2
-
     - uses: fwal/setup-swift@v1
       with:
         swift-version: ${{ matrix.swift }}
+    - uses: ruby/setup-ruby@v1
 
     - name: Test CLI Package
       run: swift test --enable-test-discovery

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -20,6 +20,9 @@ jobs:
       with:
         swift-version: ${{ matrix.swift }}
     - uses: ruby/setup-ruby@v1
+      with:
+        ruby-version: 2.6 # Not needed with a .ruby-version file
+        bundler-cache: true # runs 'bundle install' and caches installed gems automatically
 
     - name: Test CLI Package
       run: swift test -v

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -8,10 +8,17 @@ env:
 
 jobs:
   build:
-    runs-on: macos-latest
-
+    name: Build with ${{ matrix.swift }} on ${{ matrix.os }}
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [macos-latest, ubuntu-latest]
+        swift: ["5.3"]
     steps:
     - uses: actions/checkout@v2
+    - uses: fwal/setup-swift@v1
+      with:
+        swift-version: ${{ matrix.swift }}
 
     - name: Test CLI Package
       run: swift test -v
@@ -20,6 +27,7 @@ jobs:
       run: make install
       
     - name: Test VariantsCore
+      if: matrix.os == 'macos-latest' # xcode available on macos only
       run: xcodebuild test -scheme VariantsCore
 
     - name: Bundle install

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -20,8 +20,9 @@ jobs:
       with:
         swift-version: ${{ matrix.swift }}
     - uses: ruby/setup-ruby@v1
+      if: matrix.os != 'macos-latest'
       with:
-        ruby-version: 2.6 # Not needed with a .ruby-version file
+        ruby-version: 2.7 # Not needed with a .ruby-version file
         bundler-cache: true # runs 'bundle install' and caches installed gems automatically
 
     - name: Test CLI Package

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -8,7 +8,7 @@ env:
 
 jobs:
   build:
-    name: Build with ${{ matrix.swift }} on ${{ matrix.os }}
+    name: Build with Swift ${{ matrix.swift }} on ${{ matrix.os }}
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
@@ -23,7 +23,7 @@ jobs:
       if: matrix.os != 'macos-latest'
       with:
         ruby-version: 2.7 # Not needed with a .ruby-version file
-        bundler-cache: true # runs 'bundle install' and caches installed gems automatically
+        # bundler-cache: true # runs 'bundle install' and caches installed gems automatically
 
     - name: Test CLI Package
       run: swift test -v
@@ -39,9 +39,11 @@ jobs:
       run: bundle install
       
     - name: Code coverage report
+      if: matrix.os == 'macos-latest' # on macos only
       run: bundle exec slather coverage
 
     - uses: codecov/codecov-action@v1
+      if: matrix.os == 'macos-latest' # on macos only
       with:
         token: ${{ secrets.CODECOV_TOKEN }}
         file: ./test-reports/cobertura.xml

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -19,6 +19,7 @@ jobs:
     - uses: fwal/setup-swift@v1
       with:
         swift-version: ${{ matrix.swift }}
+    - uses: ruby/setup-ruby@v1
 
     - name: Test CLI Package
       run: swift test -v

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -13,7 +13,7 @@ jobs:
     strategy:
       matrix:
         os: [macos-latest, ubuntu-latest]
-        swift: ["5.3"]
+        swift: ["5.5"]
     steps:
     - uses: actions/checkout@v2
     - uses: fwal/setup-swift@v1

--- a/Gemfile
+++ b/Gemfile
@@ -1,4 +1,4 @@
 source 'https://rubygems.org'
 
-gem 'slather'
-gem 'nokogiri', '1.12.5'
+gem "slather"
+gem "nokogiri", ">= 1.13.2"

--- a/Gemfile
+++ b/Gemfile
@@ -2,3 +2,4 @@ source 'https://rubygems.org'
 
 gem "slather"
 gem "nokogiri", ">= 1.13.2"
+gem  "racc", "1.5.2"

--- a/Gemfile
+++ b/Gemfile
@@ -2,4 +2,3 @@ source 'https://rubygems.org'
 
 gem "slather"
 gem "nokogiri", ">= 1.13.2"
-gem  "racc", "1.5.2"

--- a/Makefile
+++ b/Makefile
@@ -38,19 +38,19 @@ variants: $(SOURCES)
 
 .PHONY: install
 install: variants
-	@install -d "$(bindir)" "$(libdir)"
-	@install "$(BUILDDIR)/release/variants" "$(bindir)"
-	@mkdir -p "$(libdir)/variants"
-	@cp -R "$(TEMPLATES)" "$(libdir)/variants/"
-	@cp -R "$(UTILS)" "$(libdir)/variants/"
+	@install -d $(bindir) $(libdir)
+	@install "$(BUILDDIR)/release/variants" $(bindir)
+	@mkdir -p $(libdir)/variants
+	@cp -R "$(TEMPLATES)" $(libdir)/variants/
+	@cp -R "$(UTILS)" $(libdir)/variants/
 
 .PHONY: ci
 ci:
-	@install -d "$(bindir)" "$(libdir)"
-	@install "$(CIBUILDDIR)/release/variants" "$(bindir)"
-	@mkdir -p "$(libdir)/variants"
-	@cp -R "$(TEMPLATES)" "$(libdir)/variants/"
-	@cp -R "$(UTILS)" "$(libdir)/variants/"
+	@install -d $(bindir) $(libdir)
+	@install "$(CIBUILDDIR)/release/variants" $(bindir)
+	@mkdir -p $(libdir)/variants
+	@cp -R "$(TEMPLATES)" $(libdir)/variants/
+	@cp -R "$(UTILS)" $(libdir)/variants/
 
 .PHONY: pre-ci
 pre-ci: variants
@@ -58,8 +58,8 @@ pre-ci: variants
 
 .PHONY: uninstall
 uninstall:
-	@rm -rf "$(bindir)/variants"
-	@rm -rf "$(libdir)/variants"
+	@rm -rf $(bindir)/variants
+	@rm -rf $(libdir)/variants
 
 .PHONY: clean
 distclean:

--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
 SHELL = /bin/bash
 
-prefix ?= /usr/local
+prefix ?= .local
 bindir ?= $(prefix)/bin
 libdir ?= $(prefix)/lib
 srcdir = Sources
@@ -15,6 +15,12 @@ TEMPLATES = $(templatesdir)
 UTILS = $(utilsdir)
 
 .DEFAULT_GOAL = all
+
+ifeq ($(OS),Windows_NT)
+	detected_OS := Windows
+else
+	detected_OS := $(shell uname)
+endif
 
 .PHONY: all
 all: variants
@@ -65,7 +71,9 @@ prepare_for_test:
 .PHONY: test
 test: prepare_for_test
 	@swift test
-	@xcodebuild test -scheme VariantsCore 
+ifeq ($(detected_OS),Darwin) # Mac OSX only
+	@xcodebuild test -scheme VariantsCore
+endif
 
 .PHONY: coverage
 coverage: test

--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,17 @@
 SHELL = /bin/bash
 
-prefix ?= .local
+ifeq ($(OS),Windows_NT)
+	detected_OS := Windows
+else
+	detected_OS := $(shell uname)
+endif
+
+ifeq ($(detected_OS),Linux) # Linux only
+	prefix ?= .local
+else
+	prefix ?= /usr/local
+endif
+
 bindir ?= $(prefix)/bin
 libdir ?= $(prefix)/lib
 srcdir = Sources
@@ -15,12 +26,6 @@ TEMPLATES = $(templatesdir)
 UTILS = $(utilsdir)
 
 .DEFAULT_GOAL = all
-
-ifeq ($(OS),Windows_NT)
-	detected_OS := Windows
-else
-	detected_OS := $(shell uname)
-endif
 
 .PHONY: all
 all: variants

--- a/Makefile
+++ b/Makefile
@@ -7,7 +7,7 @@ else
 endif
 
 ifeq ($(detected_OS),Linux) # Linux only
-	prefix ?= .local
+	prefix ?= ~/.local
 else
 	prefix ?= /usr/local
 endif

--- a/README.md
+++ b/README.md
@@ -55,6 +55,8 @@ brew install backbase/m/variants
 
 ### Make
 
+> For **Linux**, make sure you have [Swift installed](https://www.swift.org/getting-started/#installing-swift) by running `swift --version`
+
 ```sh
 git clone https://github.com/backbase/variants.git
 cd variants

--- a/README.md
+++ b/README.md
@@ -55,7 +55,7 @@ brew install backbase/m/variants
 
 ### Make
 
-> For **Linux**, make sure you have [Swift installed](https://www.swift.org/getting-started/#installing-swift) by running `swift --version`
+> For **Linux**, make sure you have [Swift installed](https://www.swift.org/getting-started/#installing-swift) by running `swift --version` 
 
 ```sh
 git clone https://github.com/backbase/variants.git

--- a/Tests/VariantsCoreTests/SecretsFactoryTests.swift
+++ b/Tests/VariantsCoreTests/SecretsFactoryTests.swift
@@ -62,7 +62,13 @@ class SecretsFactoryTests: XCTestCase {
         XCTAssertEqual(try String(contentsOfFile: variantsFile), variantsSwiftContent)
     }
     
+    func testUtilsDirectory_pathExists() {
+        XCTAssertEqual(try UtilsDirectory().path.exists, true)
+        XCTAssertEqual(try TemplateDirectory().path.exists, true)
+    }
+    
     static var allTests = [
-        ("testRender_correctData", testRender_noSecrets)
+        ("testRender_correctData", testRender_noSecrets),
+        ("testUtilsDirectory_pathExists", testUtilsDirectory_pathExists)
     ]
 }

--- a/Tests/VariantsTests/VariantsTests.swift
+++ b/Tests/VariantsTests/VariantsTests.swift
@@ -51,7 +51,7 @@ class CLIExecutor {
             return nil
         }
 
-        let fooBinary = productsDirectory.appendingPathComponent("Variants")
+        let fooBinary = productsDirectory.appendingPathComponent("variants")
 
         let process = Process()
         process.executableURL = fooBinary

